### PR TITLE
FragmentInjector now injects from parent fragment when present

### LINKED ISSUE #53: Fragments should use the parent fragment's injector as the parent when possible
As of Android API 17 (Jelly Bean MR1), a `Fragment` can contain another `Fragment`. It makes sense for cases where a `FragmentInjector` has a `parentFragment` that it's injector is used as the parent instead of the `Activity`.

The specific issue I encountered was a parent `Fragment` was providing additional bindings in `provideOverridingModule` that were not present in the child `Fragment`.

### PR DESCRIPTION
**Pull request by coreydowning**
Fixes #53. Resolved an issue from original PR with overriding the `Fragment` binding.

### CODE REVIEW

File: kodein-android/src/main/java/com/github/salomonbrys/kodein/android/KodeinAndroidComponents.kt [hunk 1]
```diff
@@ -302,14 +308,16 @@ interface FragmentInjector : AndroidInjector<Fragment, AndroidScope<Fragment>> {
             activity?.let {
                 bindErased<LayoutInflater>(ACTIVITY_LAYOUT_INFLATER, overrides = true) with erasedInstance(activity.layoutInflater)
             }
-            bindErased<Fragment>() with erasedInstance(kodeinComponent)
+            bindErased<Fragment>(overrides = parentFragment != null) with erasedInstance(kodeinComponent)
```
Comment (eygraber) [inline kodein-android/src/main/java/com/github/salomonbrys/kodein/android/KodeinAndroidComponents.kt]: There should probably be a tagged binding to retrieve the parent if it is not null.

**Comment by eygraber** [comment]
@coreydowning Is there any reason for not doing this for `SupportFragment` as well?

**Comment by coreydowning** [comment]
I'd rather delete `KodeinSupportFragment` outright. I am making some other changes to this PR so I will include this change.

I do not agree with providing a tagged binding for the parent--it's already available on the `Fragment`.

**Comment by eygraber** [comment]
Why would you delete `KodeinSupportFragment`? I use it extensively in my app.

**Comment by coreydowning** [comment]
I mostly wish Google would phase out support library functionality that should be deprecated because new applications do not need to support 2.x devices. I also would like to see them continue the trend of transparently using the "correct" version of classes automatically like they do with AppCompat view inflation.

I'm not actually going to delete something that's already supported.

**Comment by eygraber** [comment]
It would be nice if that were the case, but it won't be for a long time. It doesn't just support 2.x devices; the support library adds nested fragments for APIs lower than 17, etc...

I also wouldn't use `onAttach` / `onDetach` because it can lead to lots of memory being used unnecessarily in situations where `Fragment` is destroyed but not detached.

For example, if you use `replace` in a `FragmentTransaction` then `onDetach` won't get called until the `Fragment` is explicitly removed or the `Activity` is finished. `onCreate` / `onDestroy` offers more predictable call behavior.

**Comment by coreydowning** [comment]
My goal with the move to `onAttach` and `onDetach` was to make Kodein work somewhat similarly to Dagger 2, which has `Fragment` injection occur in `onAttach`. It's been a long while since I've dug into the specifics of these early lifecycle methods so this is my only argument.

**Comment by eygraber** [comment]
I'd assume Dagger 2 works that way because it's the earliest logical place to put it. In the Android components PR I had it that way myself initially, but it caused big memory issues in my app that I was testing it, so I changed it to `onCreate` and `onDestroy`.

I don't know if the Dagger 2 team discovered that issue yet, and if they did they might've decided that the ability to have injections in `onAttach` was more important.

My view is that there's not much gain in having injections available in `onAttach` because `onCreate` happens right afterwards. Since it also has the memory problem, I'd vote to stick with `onCreate` / `onDestroy`.

**Comment by coreydowning** [comment]
I'll go ahead and move `Fragment` injection back to `onCreate`/`onDestroy` and squash this branch down. I'd also like to see some of the active PRs merged in, any input on these @SalomonBrys ?

**Comment by SalomonBrys** [comment]
Looks good.

I would like to apologize for the time it took me to review this.
I've had a combination of health issue, wedding prep and lots of work at the office ;)

Thanks a lot for the PR :smile:

**Comment by eygraber** [comment]
@coreydowning I've been thinking about `onAttach`/`onDetach` every once in a while. 

Fragments with `setRetainInstance(true)` don't call `onDestroy`, but they do call `onDetach`. So `onAttach` / `onDetach` are guaranteed for all fragments. 

There's still the memory usage issue though, but lately I've been thinking that having the user handle that is acceptable if it means that retained fragments will work. 

What do you think?

**Comment by coreydowning** [comment]
Yeah this looks like a good reason to move to `onAttach`/`onDetach`.

**Pull request merged**